google/cloud/storage/retry.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from google.api_core import exceptions
+from google.api_core import retry
+
+import json
+
+
+_RETRYABLE_REASONS = frozenset(
+    ["rateLimitExceeded", "backendError", "internalError", "badGateway"]
+)
+
+_UNSTRUCTURED_RETRYABLE_TYPES = (
+    exceptions.TooManyRequests,
+    exceptions.InternalServerError,
+    exceptions.BadGateway,
+)
+
+
+# FIXME: needs to be brought in line with doc outlining all retriable error codes
+# FIXME: add tests once above is done
+def _should_retry(exc):
+    """Predicate for determining when to retry."""
+    if not hasattr(exc, "errors"):
+        return False
+
+    if len(exc.errors) == 0:
+        # Check for unstructured error returns, e.g. from GFE
```

Comment:
unabbreviate for public understanding?

Reply:
Removed (this was part of google-api-core; my hands-on testing suggests the entire section is unneeded for GCS)